include/aspect/particle/property/crystal_preferred_orientation.h

```diff
@@ -255,7 +222,93 @@ namespace aspect
           unsigned int
           get_number_of_minerals();
 
+          /**
+           * return a reference to the value in the data array representing the
+           * deformation type.
+           */
+          inline
+          double &ref_deformation_type(const unsigned int cpo_data_position,
```

Comment:
Already better, but returning a reference is a somewhat dangerous design, because it is easy to accidentally change a value that you just want to read (plus the naming with ref_ ... is awkward). I would prefer you split all of these into get_ set_ functions like you do below.
I can see two reasons why you did it this way: 1. less function. 2. the construct below where you do `ref_function(...) *= some number` which saves you one function call by having the reference. However, for both cases I would still prefer to have separate getter/setter functions like you do for the rotation matrix. The number of functions does not matter much, since it is just a single line per function. And the performance loss by having to call the get_ function before multiplying the value and then calling the set_ function is actually very small (compared to creating a lot of vectors like you did before). Consider the following:
- A multiplication or addition costs around one clock cycle. You do maybe 10 of those in the function to determine the index plus you inline the function so the function call is nearly free.
- Accessing a variable that is already in L1 cache of the processor costs at most a few cycles (<10). This happens when you call the set_ function after calling the get_ function.
- Creating a variable on the stack (like a double or a single Tensor) costs less than 5 cycles (probably more like 1).
- So the total overhead of having separate get_/set_ functions is likely less than 20 clock cycles.
- Creating a single variable on the heap (like a std::vector) requires calling a complex algorithm that finds the necessary space, and on average costs around 90 clock cycles. 

So unless you have measured that this is indeed a performance bottleneck I would prefer an interface of the following form for all the function below:
```
inline
double get_deformation_type (... like before) const;

inline
void set_deformation_type (... like before ..., const double deformation_type) const;
```